types.go

```diff
@@ -89,11 +89,14 @@ type cfgString struct {
 	s string
 }
 
+type cfgSubFlex struct{ cfgSub }
+
 type cfgSub struct {
 	c *Config
 }
 
 type cfgNil struct{ cfgPrimitive }
+type cfgEmpty struct{ cfgPrimitive }
```

Comment:
what does differentiate the to? When to use `cfgNil` and when `cfgEmpty`?

Reply:
`cfgEmpty` is different then `cfgNil` in that it means that the value for the field was empty, not that it was set to nil. `cfgEmpty` is only used internally to reify and unpack and not returned from the configuration parser.

It allows `unpackWith` to understand that the field was empty and not nil. `cfgEmpty` returns error `ErrEmpty` which `unpackWith` understands as an allowed error that it can accept, ignore and continue on.

By changing the code back to `cfgNil` 2 tests fail with ErrTypeMismatch. This is because unpackWith is trying to unpack a value that was empty from the config, being that the code now creates values for all Arrays, Slices, Maps, and Structs so that any nested `InitDefaults` will be performed.

`unpackWith` could change to ignore the `ErrTypeMismatch` but I thought it was better to not do that so actual type errors are not ignored.